File: c++/3535_UnitConversionII.cpp
// Problem 3535: Unit Conversion II
// https://leetcode.com/problems/unit-conversion-ii/

#include <vector>
#include <functional>
#include <list>
using namespace std;

class Solution
{
public:
    static const int MOD = 1e9 + 7;
    vector<int> queryConversions(vector<vector<int>> &conversions, vector<vector<int>> &queries)
    {
        int n = conversions.size();
        // 設定答案，預設沒資料都是 -1
        vector<int> baseUnitConv(n + 1, -1);
        // 第 0 筆資料為 1 分 "第 0 筆資料單位"
        baseUnitConv[0] = 1;

        // 整理反向 map
        unordered_map<int, list<pair<int, int>>> reverseMap;
        for (const auto &conversion : conversions)
        {
            int source = conversion[0], target = conversion[1], factor = conversion[2];
            reverseMap[target].emplace_back(source, factor);
        }

        // 設定 dfs function
        function<void(int, int, int)> dfs = [&](int source, int target, int factor)
        {
            if (baseUnitConv[source] == -1)
            {
                // 假如 source 還沒設定，使用 reverseMap 往前設定
                for (const auto &prev : reverseMap[source])
                {
                    // 前一個節點使用 dfs 尋找並設定
                    dfs(prev.first, source, prev.second);
                    // 直到 source 確定設定完畢
                    if (baseUnitConv[source] != -1)
                    {
                        break;
                    }
                }
            }

            // 目標單位設定成 來源乘以倍數
            long long conv = baseUnitConv[source] % MOD;
            conv = (conv * factor % MOD) % MOD;
            baseUnitConv[target] = static_cast<int>(conv);
        };

        // 跑迴圈處理每個 conversion
        for (const auto &conversion : conversions)
        {
            int source = conversion[0], target = conversion[1], factor = conversion[2];
            dfs(source, target, factor);
        }

        // 前面段與 Unit Conversion I 相同
        // 跑迴圈依題意比較每組 a, b 的值並加入 queryConv
        vector<int> queryConv;
        for (const auto &query : queries)
        {
            int unitA = query[0], unitB = query[1];
            // 使用模反元素來計算轉換結果
            long long conv = baseUnitConv[unitB];
            conv = (conv * modInverse(baseUnitConv[unitA])) % MOD;
            queryConv.push_back(static_cast<int>(conv % MOD));
        }

        // 回傳答案
        return queryConv;
    }

private:
    // 計算模反元素
    int modInverse(long num)
    {
        int result = 1;
        int exponent = MOD - 2;
        while (exponent > 0)
        {
            if (exponent % 2 == 1)
            {
                result = (result * num) % MOD;
            }
            num = (num * num) % MOD;
            exponent /= 2;
        }
        return int(result);
    }
};

File: python/3535_unit_conversion_ii.py
# Problem 3535: Unit Conversion II
# https://leetcode.com/problems/unit-conversion-ii/

from collections import defaultdict
from typing import List

class Solution:
    def query_conversions(self, conversions: List[List[int]], queries: List[List[int]]) -> List[int]:
        # 設定空陣列為答案
        base_unit_conv = [None for _ in range(len(conversions)+1)]
        # 第 0 筆資料為 1 分 "第 0 筆資料單位"
        base_unit_conv[0] = 1

        MOD = 10**9 + 7

        # 整理反向 dict，可以回推找資料
        reverse_map = defaultdict(list)
        for source, target, factor in conversions:
            reverse_map[target].append((source, factor)) 

        # 設定 dfs function
        def process_source_conv(source, target, factor):
            if base_unit_conv[source] is None:
                # 假如 source 還沒設定，使用 reverse_map 往前設定
                for prev, prev_factor in reverse_map[source]:
                    # 前一個節點使用 process_source_conv 尋找並設定
                    process_source_conv(prev, source, prev_factor)
                    # 直到 source 確定設定完畢
                    if base_unit_conv[source] is not None:
                        break

            # 目標單位設定成 來源乘以倍數
            base_unit_conv[target] = (base_unit_conv[source]*factor)%MOD
            return

        # 跑迴圈處理每個 conversions
        for source, target, factor in conversions:
            process_source_conv(source, target, factor)

        # 前面段與 Unit Conversion I 相同
        # 跑迴圈依題意比較每組 a, b 的值並加入 query_conv
        query_conv = []
        for unit_a, unit_b in queries:
            query_conv.append(base_unit_conv[unit_b] * pow(base_unit_conv[unit_a], MOD - 2, MOD) % MOD)

        # 回傳答案
        return query_conv


